File: src/cpp/MultipleInheritance.cpp
//: src/cpp/MultipleInheritance.cpp
// Combining disjoint types using MI
#include <iostream>
using namespace std;

class Person {
    public:
    void consume() { cout << "eating pizza" << endl; }
};

class Robot {
    public:
    void charge() { cout << "charging" << endl; }
};

class Base {
    public:
    virtual void eat() = 0;  // Pure virtual == abstract
};
//---
class Person2: public Base, private Person {
    public:
    void eat() override { consume(); }
};

class Robot2: public Base, private Robot {
    public:
    void eat() override { charge(); }
};

void nourish(Base* base) {
    base->eat();
}

int main() {
    Base* subjects[] = { new Person2(), new Robot2() };
    for(Base* subject: subjects)
      nourish(subject);
}

File: src/python/multiple_inheritance.py
#: src/python/multiple_inheritance.py
from base import Base
from disjoint_types import Person, Robot

class Person2(Person, Base):
    def eat(self): super().consume()

class Robot2(Robot, Base):
    def eat(self): super().charge()

def nourish(base: Base):
    base.eat()

if __name__ == '__main__':
    for subject in [Person2(), Robot2()]:
        nourish(subject)
